keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,570 @@
+  Pods.
+
+### Non-Goals
+
+- Remove Pods once they have been accounted for.
+
+## Proposal
+
+The Job controller creates Pods with a finalizer to prevent finished Pods from
+being removed by the garbage collector. The Job controller removes the finalizer
+from the finished Pods once it has accounted for them. In subsequent Job syncs,
+the controller ignores finished Pods that don't have the finalizer.
+
+### Notes/Constraints/Caveats (Optional)
+
+Due to the lack of support of atomic changes across kubernetes objects, an
+intermediate state is necessary. Before removing the Pod finalizers,
+the controller adds finished Pods to a list in the Job status. After removing
+the Pod finalizers, the controller clears the list and updates the counters.
+
+### Risks and Mitigations
+
+#### New API calls
+
+The new algorithm introduces new API calls:
+- one per Pod lifecycle to remove finalizers
+- one for each Job sync, to track the intermediate state.
+
+Note that there no new calls in the Pod creation path.
+
+On the other hand, to update the Job status once a Pod finishes, we need a total
+2 new API calls compared to the legacy algorithm. If more than one Pod finish at
+a given time, we add `n + 1` API calls, where `n` is the number finished Pods.
+
+This increase in API calls is justified for the following reasons:
+- The legacy tracking cannot handle a big number of terminated Pods, across
+  any number of Jobs, at a given point.
+- In the entirety of its lifecycle, a Pod requires at least 8 API calls,
+  including status updates and events.
+  
+However, in order to prevent Jobs with big number of Pods from starving Jobs
+with fewer Pods, the Job controller might skip status updates until enough
+Pods have accumulated or enough time has passed. See [Algorithm](#algorithm)
+below for more details.
+   
+#### Bigger Job status
+
+Job status can temporarily grow if too many Pods finish at the same time or
+if the Job controller is down for some time. In this case, we do partial
+status updates. See [Algorithm](#algorithm) below.
+
+#### Unprotected Job status endpoint
+
+Changes in the status not produced by the Job controller in
+kube-controller-manager could affect the Job tracking. Cluster administrators
+should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedTerminatedPods holds UIDs of Pods that have finished but
+    // haven't been accounted in the counters.
+    // If nil, Job tracking doesn't make use of this structure.
+    // +optional
+    UncountedTerminatedPods *UncountedTerminatedPods
+}
+
+// UncountedTerminatedPods holds UIDs of Pods that have finished but haven't
+// been accounted in Job status counters.
+type UncountedTerminatedPods struct {
```

Comment:
I'm not in favor of this name, but we can work it out during API review and update the proposal afterwards.

Reply:
Ack